### PR TITLE
[sonic_installer] Adding support for ARM Uboot firmware and portstat fix

### PR DESCRIPTION
**Pull request by antony-rheneus**
[sonic_installer] Adding support for ARM by introducing bootloader type 
grub vs uboot
For uboot fw_setenv/fw_printenv commands are utilized here
fw_env.cfg config needs to be updated per platform

[portstat] division by zero error fix when speed is < 1000 (1G)

Signed-off-by: Antony Rheneus <arheneus@marvell.com>

**- What I did**

**- How I did it**

**- How to verify it**

**- Previous command output (if the output of a command-line utility has changed)**

**- New command output (if the output of a command-line utility has changed)**

### CODE REVIEW

**Comment by antony-rheneus** [comment]
This is a fork of PR https://github.com/Azure/sonic-utilities/pull/731 which was closed due to merge conflict issues.

**Comment by antony-rheneus** [comment]
Can you please retest, not able to figure out the Failures.

File: scripts/portstat [hunk 1]
```diff
@@ -109,7 +109,7 @@ class Portstat(object):
         if speed is None:
             speed = PORT_RATE
         else:
-            speed = int(speed)/1000
+            speed = float(speed)/float(1000)
```
Comment (lguohan) [inline scripts/portstat]: can you not change this? I'm not sure if there is requirement to set speed to integer

**Comment by lguohan** [comment]
retest this please

File: sonic_installer/main.py [hunk 2]
```diff
@@ -106,9 +111,15 @@ def set_default_image(image):
     if get_running_image_type() == IMAGE_TYPE_ABOOT:
         image_path = aboot_image_path(image)
         aboot_boot_config_set(SWI=image_path, SWI_DEFAULT=image_path)
-    else:
+    elif BOOTLOADER == BOOTLOADER_TYPE_GRUB:
         command = 'grub-set-default --boot-directory=' + HOST_PATH + ' ' + str(images.index(image))
         run_command(command)
+    elif BOOTLOADER == BOOTLOADER_TYPE_UBOOT:
+        if image in images[0]:
+            run_command('/usr/bin/fw_setenv boot_next "run sonic_image_1"')
```
Comment (LuiSzee) [inline sonic_installer/main.py:119]: Using fw_printenv/fw_setenv to get/set a uboot environment variable is platform specific.
I think platform specific code shouldn't appear here. Can you move it to platform directory.

File: sonic_installer/main.py [hunk 3]
```diff
@@ -156,14 +167,25 @@ def get_installed_images():
...
-    else:
+    elif BOOTLOADER == BOOTLOADER_TYPE_GRUB:
         config = open(HOST_PATH + '/grub/grub.cfg', 'r')
         for line in config:
             if line.startswith('menuentry'):
                 image = line.split()[1].strip("'")
                 if IMAGE_PREFIX in image:
                     images.append(image)
         config.close()
+    elif BOOTLOADER == BOOTLOADER_TYPE_UBOOT:
```
Comment (LuiSzee) [inline sonic_installer/main.py:178]: It seems that you can only support 2 images.

File: sonic_installer/main.py [hunk 4]
```diff
@@ -156,14 +167,25 @@ def get_installed_images():
...
+        proc = subprocess.Popen("/usr/bin/fw_printenv -n sonic_version_1", shell=True, stdout=subprocess.PIPE)
```
Comment (LuiSzee) [inline sonic_installer/main.py:179]: Can you move it to platform directory?

File: sonic_installer/main.py [hunk 5]
```diff
@@ -192,6 +214,16 @@ def get_next_image():
             else:
                 next_image_index = 0
         next_image = images[next_image_index]
+    elif BOOTLOADER == BOOTLOADER_TYPE_UBOOT:
+        images = get_installed_images()
+        proc = subprocess.Popen("/usr/bin/fw_printenv -n boot_next", shell=True, stdout=subprocess.PIPE)
```
Comment (LuiSzee) [inline sonic_installer/main.py:219]: Can you move it to platform directory?

File: sonic_installer/main.py [hunk 6]
```diff
@@ -226,6 +258,19 @@ def remove_image(image):
 
         run_command('grub-set-default --boot-directory=' + HOST_PATH + ' 0')
         click.echo('Image removed')
+    elif BOOTLOADER == BOOTLOADER_TYPE_UBOOT:
+        click.echo('Updating next boot ...')
+        images = get_installed_images()
+        if image in images[0]:
+            run_command('/usr/bin/fw_setenv boot_next "run sonic_image_2"')
```
Comment (LuiSzee) [inline sonic_installer/main.py:265]: the code to delete image is also platform specific

File: sonic_installer/main.py [hunk 7]
```diff
@@ -410,9 +456,14 @@ def set_next_boot(image):
     if get_running_image_type() == IMAGE_TYPE_ABOOT:
         image_path = aboot_image_path(image)
         aboot_boot_config_set(SWI=image_path)
-    else:
+    elif BOOTLOADER == BOOTLOADER_TYPE_GRUB:
         command = 'grub-reboot --boot-directory=' + HOST_PATH + ' ' + str(images.index(image))
         run_command(command)
+    elif BOOTLOADER == BOOTLOADER_TYPE_UBOOT:
+        if image in images[0]:
+            run_command('/usr/bin/fw_setenv boot_once "run sonic_image_1"')
```
Comment (LuiSzee) [inline sonic_installer/main.py:464]: this code is platform specific

**Comment by antony-rheneus** [comment]
> Some code in soinc_installer/main.py is platform specific. please handle it

@LuiSzee, Any plans to move "grub" related code to platform specific path. Do you have design for such platform specific bootloader env change.
If so please point it, so that I can follow the same for "uboot" related code.

**Pull request merged**